pkg/dashboard/functiontemplates/githubFunctionTemplateFetcher.go

```diff
@@ -0,0 +1,192 @@
+
+	return &GithubFunctionTemplateFetcher{
+		Repository:                   repository,
+		Owner:                        owner,
+		Branch:                       branch,
+		githubAPIClient:              client,
+		supportedSourceTypesSuffixes: supportedSourceTypesSuffixes,
+	}, nil
+}
+
+func (gftf *GithubFunctionTemplateFetcher) Fetch() ([]FunctionTemplate, error) {
+	var functionTemplates []FunctionTemplate
+
+	// get sha of root of source tree
+	treeSha, err := gftf.getSourceTreeSha()
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get source tree sha")
+	}
+
+	// get templates from source tree sha
+	functionTemplates, err = gftf.getTemplatesFromGithubSHA(treeSha, "")
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get templates from source tree sha")
+	}
+
+	return functionTemplates, nil
+}
+
+func (gftf *GithubFunctionTemplateFetcher) getTemplatesFromGithubSHA(treeSha string, upperDirName string) ([]FunctionTemplate, error) {
+	var functionTemplates []FunctionTemplate
+
+	// get subdir items from github sha
+	// recursive set to false because when set to true it may not give all items in dir (https://developer.github.com/v3/git/trees/#get-a-tree-recursively)
+	tree, _, err := gftf.githubAPIClient.Git.GetTree(context.TODO(), gftf.Owner, gftf.Repository, treeSha, false)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get source tree with GetTree go-github function")
+	}
+
+	// add template if there is one in current dir
+	currentDirTemplate, err := gftf.getTemplateFromDir(tree.Entries, upperDirName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to template from dir")
+	}
+
+	// add found template to function templates
+	if currentDirTemplate != nil {
+		functionTemplates = append(functionTemplates, *currentDirTemplate)
+	}
+
+	// search recursively in other entries (items in current dir) which are dirs
+	for _, entry := range tree.Entries {
+		if *entry.Type == "tree" {
+			// get subdir templates
+			subdirTemplates, err := gftf.getTemplatesFromGithubSHA(*entry.SHA, *entry.Path)
+			if err != nil {
+				return nil, errors.Wrap(err, "Failed to get templates from sub directory")
+			}
+
+			functionTemplates = append(functionTemplates, subdirTemplates...)
+		}
+	}
+
+	return functionTemplates, nil
+}
+
+func (gftf *GithubFunctionTemplateFetcher) getTemplateFromDir(dir []github.TreeEntry, upperDirName string) (*FunctionTemplate, error) {
+	functionTemplate := FunctionTemplate{}
+
+	// add dir name as function's Name
+	functionTemplate.Name = upperDirName
+
+	if sourceFile, err := gftf.getFirstSourceFile(dir, gftf.supportedSourceTypesSuffixes); sourceFile != nil {
+		functionTemplate.SourceCode = *sourceFile
+	} else if err != nil {
+		return nil, errors.Wrap(err, "Failed to get and process source file")
+	}
+
+	// get function.yaml - error if failed to get its content although it exists
+	if file, err := gftf.getFileFromTreeEntries(dir, "function.yaml"); err != nil {
```

Comment:
1. make functionTemplate.FunctionConfig a pointer (*functionTemplate.FunctionConfig)
2. the if / else thing is hard to follow - too many things to "hold in your head". refactor the logic here to:
```
functionTemplate.FunctionConfig, err = getFunctionConfigFromTreeEntries()
if err {
	...
}

// if we got a functionconfig we're done
if functionTemplate.FunctionConfig != nil {
	return
}

// get template and values - either BOTH exist or NONE exist. Otherwise err is set
functionYAMLTemplateContents, functionYAMLValuesContents, err := getFunctionYAMLTemplateAndValuesFromTreeEntries()
if err {
	...
}

// if values is set, parse
if functionYAMLValuesContents != "" {
	
	// parse to functionTemplate.FunctionConfigValues
}
```